domain/application/state/application.go

```diff
@@ -875,33 +871,160 @@ INSERT INTO k8s_service (*) VALUES ($cloudService.*)
-		}
-		serviceInfo.ProviderID = providerID
-		uuid, err := uuid.NewUUID()
-		if err != nil {
+		if err := st.insertCloudServiceAddresses(ctx, tx, applicationName, linkLayerDeviceUUID.String(), sAddrs); err != nil {
 			return errors.Capture(err)
 		}
-		serviceInfo.UUID = uuid.String()
-		return tx.Query(ctx, insertStmt, serviceInfo).Run()
+		return nil
 	})
 	if err != nil {
 		return errors.Errorf("updating cloud service for application %q: %w", applicationName, err)
 	}
 	return nil
 }
 
+func (st *State) insertCloudServiceNetNode(ctx context.Context, tx *sqlair.TX, serviceInfo cloudService, applicationName string) (uuid.UUID, error) {
+	createNodeStmt, err := st.Prepare(`
+INSERT INTO net_node (uuid) VALUES ($cloudService.net_node_uuid)
+`, serviceInfo)
+	if err != nil {
+		return uuid.UUID{}, errors.Capture(err)
+	}
+
+	nodeUUID, err := uuid.NewUUID()
+	if err != nil {
+		return uuid.UUID{}, errors.Capture(err)
+	}
+	serviceInfo.NetNodeUUID = nodeUUID.String()
+	if err := tx.Query(ctx, createNodeStmt, serviceInfo).Run(); err != nil {
+		return uuid.UUID{}, errors.Errorf("creating cloud service net node for application %q: %w", applicationName, err)
+	}
+	return nodeUUID, nil
+}
+
+func (st *State) insertCloudServiceDevice(ctx context.Context, tx *sqlair.TX, applicationName string, netNodeUUID string) (uuid.UUID, error) {
+	// For cloud services, the device is a placeholder without
+	// a MAC address and once inserted, not updated. It just exists
+	// to tie the address to the net node corresponding to the
+	// cloud service.
+	devUUID, err := uuid.NewUUID()
+	if err != nil {
+		return uuid.UUID{}, errors.Capture(err)
+	}
+	cloudServiceDeviceInfo := cloudServiceDevice{
+		UUID:              devUUID.String(),
+		Name:              fmt.Sprintf("placeholder for %q cloud service", applicationName),
+		DeviceTypeID:      int(linklayerdevice.DeviceTypeUnknown),
+		VirtualPortTypeID: int(linklayerdevice.NonVirtualPortType),
+		NetNodeID:         netNodeUUID,
+	}
+	insertCloudServiceDeviceStmt, err := st.Prepare(`
+INSERT INTO link_layer_device (*) VALUES ($cloudServiceDevice.*)
+`, cloudServiceDeviceInfo)
+	if err != nil {
+		return uuid.UUID{}, errors.Capture(err)
+	}
+
+	if err := tx.Query(ctx, insertCloudServiceDeviceStmt, cloudServiceDeviceInfo).Run(); err != nil {
+		return uuid.UUID{}, errors.Capture(err)
+	}
+	return devUUID, nil
+}
+
+func (st *State) deleteCloudServiceAddresses(ctx context.Context, tx *sqlair.TX, appUUID coreapplication.ID, providerID string) error {
+	cloudService := cloudService{
+		ApplicationUUID: appUUID,
+		ProviderID:      providerID,
+	}
+	deleteAddressStmt, err := st.Prepare(`
+DELETE FROM ip_address
+WHERE device_uuid IN (
+    SELECT device_uuid 
+    FROM link_layer_device AS lld
+    JOIN net_node AS nn ON nn.uuid = lld.net_node_uuid
+    JOIN k8s_service AS ks ON ks.net_node_uuid = nn.uuid
+    WHERE ks.application_uuid = $cloudService.application_uuid
+    AND ks.provider_id = $cloudService.provider_id
```

Comment:
There's a lot of pre-existing code in this package but we've started to adopt the alignment like
```
    SELECT device_uuid 
    FROM   link_layer_device AS lld
    JOIN   net_node AS nn ON nn.uuid = lld.net_node_uuid
    JOIN   k8s_service AS ks ON ks.net_node_uuid = nn.uuid
    WHERE  ks.application_uuid = $cloudService.application_uuid
    AND    ks.provider_id = $cloudService.provider_id
```

Reply:
Done.